std/crypto/throughput_test.zig

```diff
@@ -1,38 +1,198 @@
+    const throughput = @floatToInt(u64, bytes / elapsed_s);
+
+    return throughput;
+}
+
+const exchanges = []Crypto{Crypto{ .ty = crypto.X25519, .name = "x25519" }};
+
+pub fn benchmarkKeyExchange(comptime DhKeyExchange: var, comptime exchange_count: comptime_int) !u64 {
+    std.debug.assert(DhKeyExchange.minimum_key_length >= DhKeyExchange.secret_length);
+
+    var in: [DhKeyExchange.minimum_key_length]u8 = undefined;
+    prng.random.bytes(in[0..]);
+
+    var out: [DhKeyExchange.minimum_key_length]u8 = undefined;
+    prng.random.bytes(out[0..]);
+
+    var offset: usize = 0;
+    var timer = try Timer.start();
+    const start = timer.lap();
+    {
+        var i: usize = 0;
+        while (i < exchange_count) : (i += 1) {
+            _ = DhKeyExchange.create(out[0..], out, in);
+        }
+    }
+    const end = timer.read();
+
+    const elapsed_s = @intToFloat(f64, end - start) / time.ns_per_s;
+    const throughput = @floatToInt(u64, exchange_count / elapsed_s);
+
+    return throughput;
+}
 
-    try stdout.print("{}: {} MiB/s\n", @typeName(HashFunction), throughput / (1 * MiB));
+fn usage() void {
+    std.debug.warn(
+        \\throughput_test [options]
+        \\
+        \\Options:
+        \\  --filter [test-name]
+        \\  --seed   [int]
+        \\  --help
+        \\
+    );
+}
+
+fn mode(comptime x: comptime_int) comptime_int {
+    return if (builtin.mode == builtin.Mode.Debug) x / 64 else x;
+}
+
+// TODO(#1358): Replace with builtin formatted padding when available.
+fn printPad(stdout: var, s: []const u8) !void {
+    var i: usize = 0;
+    while (i < 12 - s.len) : (i += 1) {
+        try stdout.print(" ");
+    }
+    try stdout.print("{}", s);
+}
+
+pub fn main() !void {
+    var stdout_file = try std.io.getStdOut();
+    var stdout_out_stream = std.io.FileOutStream.init(&stdout_file);
+    const stdout = &stdout_out_stream.stream;
+
+    var buffer: [1024]u8 = undefined;
+    var fixed = std.heap.FixedBufferAllocator.init(buffer[0..]);
+    const args = try std.os.argsAlloc(&fixed.allocator);
+
+    var filter: ?[]u8 = "";
+
+    var i: usize = 1;
+    while (i < args.len) : (i += 1) {
+        if (std.mem.eql(u8, args[i], "--seed")) {
+            i += 1;
+            if (i == args.len) {
+                usage();
+                std.os.exit(1);
+            }
+
+            const seed = try std.fmt.parseUnsigned(u32, args[i], 10);
+            prng.seed(seed);
+        } else if (std.mem.eql(u8, args[i], "--filter")) {
+            i += 1;
+            if (i == args.len) {
+                usage();
+                std.os.exit(1);
+            }
+
+            filter = args[i];
+        } else if (std.mem.eql(u8, args[i], "--help")) {
+            usage();
+            return;
+        } else {
+            usage();
+            std.os.exit(1);
+        }
+    }
+
+    inline for (hashes) |H| {
+        // TODO: Inverted early continue case here segfaults compiler. Create reduced test case.
```

Comment:
This needs a reduced test case but I couldn't get an immediately smaller one. I'll make an issue linking to this for the moment if I don't have a reduction shortly (and the issue isn't already filed).

```#0  0x0000555558b786d0 in llvm::BasicBlock::getContext() const ()
#1  0x0000555558c083e9 in llvm::BranchInst::BranchInst(llvm::BasicBlock*, llvm::BasicBlock*, llvm::Value*, llvm::Instruction*) ()
#2  0x0000555558ba3208 in LLVMBuildCondBr ()
#3  0x0000555555efd84a in ir_render_instruction(CodeGen*, IrExecutable*, IrInstruction*) [clone .isra.182] ()
#4  0x0000555555f03a0e in codegen_build(CodeGen*) ()
#5  0x0000555555e49731 in main ()
```